File: jar/src/main/kotlin/jline/lang/processes/DiscreteSampler.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.processes;

import jline.util.Pair;
import jline.util.RandomManager;
import jline.util.matrix.Matrix;
import org.apache.commons.math3.util.DoubleArray;
import org.apache.commons.math3.util.FastMath;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * A class for discrete distributions specified from the probability mass function
 */
public class DiscreteSampler extends DiscreteDistribution {

    public DiscreteSampler(Matrix p) {
        super("DiscreteSampler", 3, new Pair<Double, Double>(1.0, (double) p.length()));
        setParam(1, "p", p.columnMajorOrder().transpose());
        int n = p.length();
        Matrix x = new Matrix(1, n);
        for (int i = 0; i < n; i++) {
            x.set(i, i + 1);
        }
        setParam(2, "x", x.columnMajorOrder().transpose());
        Matrix f = p.columnMajorOrder().transpose().cumsumViaRow();
        double psum = p.elementSum();
        for (int i = 0; i < f.getNumCols(); i++) {
            f.set(i, f.get(i) / psum);
        }
        setParam(3, "f", f);
        setParam(4, "min", (int) x.elementMin());
        setParam(5, "max", (int) x.elementMax());
    }

    /**
     * Constructs a discrete distribution from a finite probability vector p at the points specified in vector x
     *
     * @param p - the probability of an item
     * @param x - the value of an item
     */
    public DiscreteSampler(Matrix p, Matrix x) {
        super("DiscreteSampler", 3, new Pair<Double, Double>(x.elementMin(), x.elementMax()));
        setParam(1, "p", p.columnMajorOrder().transpose());
        setParam(2, "x", x.columnMajorOrder().transpose());
        Matrix f = p.columnMajorOrder().transpose().cumsumViaRow();
        double psum = p.elementSum();
        for (int i = 0; i < f.getNumCols(); i++) {
            f.set(i, f.get(i) / psum);
        }
        setParam(3, "f", f);
    }

    public static void main(String[] args) {
        DiscreteSampler z = new DiscreteSampler(new Matrix("[0.2,0.5,0.3]"), new Matrix("[5,11,9]"));
        double[] S = z.sample(10000);
        int c1 = 0;
        for (int i = 0; i < S.length; i++) {
            if (S[i] == 9)
                c1++;
        }
        System.out.println((double) c1 / S.length);
    }

    @Override
    public double evalCDF(double t) {
        Matrix f = (Matrix) this.getParam(3).getValue();
        if (t >= 0 && t < f.length()) {
            return f.get((int) t);
        } else {
            return 0;
        }
    }

    public List<Double> evalPMF() {
        Matrix p = (Matrix) this.getParam(1).getValue();
        List<Double> retList = new ArrayList<>();
        for (int i = 0; i < p.getNumCols(); i++) {
            retList.add(p.get(i));
        }
        return retList;
    }

    @Override
    public Matrix evalPMF(List<Double> t) {
        Matrix p = (Matrix) this.getParam(1).getValue();
        Matrix x = (Matrix) this.getParam(2).getValue();
        List<Double> retList = new ArrayList<>();
        for (int i = 0; i < t.size(); i++) {
            int j = 0;
            while (j < x.length() && x.get(j) != t.get(i)) {
                j++;
            }
            retList.add(p.get(j));
        }
        return new Matrix(retList);
    }

    /**
     * Computes the distribution mean
     *
     * @return - the mean of the distribution
     */
    @Override
    public double getMean() {
        Matrix p = (Matrix) this.getParam(1).getValue();
        Matrix x = (Matrix) this.getParam(2).getValue();
        double mean = 0.0;
        for (int i = 0; i < p.length(); i++)
            mean += x.get(i) * p.get(i);
        return mean;
    }

    /**
     * Computes the distribution squared coefficient of variation (SCV = variance/mean^2)
     *
     * @return
     */
    @Override
    public double getSCV() {
        Matrix p = (Matrix) this.getParam(1).getValue();
        Matrix x = (Matrix) this.getParam(2).getValue();
        double e2 = 0.0;
        for (int k = 0; k < x.length(); k++) {
            e2 += FastMath.pow(x.get(k), 2) * p.get(k);
        }
        double ex = getMean();
        double var = e2 - ex * ex;
        return var / (ex * ex);
    }

    @Override
    public double getSkewness() {
        Matrix p = (Matrix) this.getParam(1).getValue();
        Matrix x = (Matrix) this.getParam(2).getValue();
        double e3 = 0.0;
        double e1 = getMean();
        double e2 = 0.0;
        for (int k = 0; k < x.length(); k++) {
            e2 += FastMath.pow(x.get(k), 2) * p.get(k);
        }

        for (int k = 0; k < x.length(); k++) {
            e3 += FastMath.pow(x.get(k), 3) * p.get(k);
        }

        return (e3 - 3 * e1 * e2 + 2 * e1 * e1 * e1) / FastMath.pow(getVar(), 1.5);
    }

    @Override
    public boolean isDisabled() {
        Matrix p = (Matrix) this.getParam(1).getValue();
        for (int i = 0; i < p.length(); i++) {
            if (Double.isNaN(p.get(i)))
                return true;
        }
        return false;
    }

    public double sample() {
        return this.sample(1, RandomManager.getThreadRandomAsRandom())[0];
    }

    @Override
    public double[] sample(int n) {
        return this.sample(n, RandomManager.getThreadRandomAsRandom());
    }

    public double[] sample(int n, Random random) {
        double[] x = ((Matrix) this.getParam(2).getValue()).toArray1D();
        double[] f = ((Matrix) this.getParam(3).getValue()).toArray1D();
        double[] ret = new double[n];
        for (int i = 0; i < n; i++) {
            double r = random.nextDouble();
            for (int q = 0; q < f.length; q++) {
                if (r <= f[q]) {
                    ret[i] = x[q];
                    break;
                }
            }
        }
        return ret;
    }
}

File: matlab/src/lang/processes/DiscreteSampler.m
classdef DiscreteSampler < DiscreteDistribution
    % A class for discrete distributions specified from the probability mass function
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.
    
    methods
        function self = DiscreteSampler(p,x)
            % SELF = DISCRETESAMPLER(P,X)
            
            % Construct a discrete distribution from a finite probability
            % vector p at the points specificied in vector x
            
            % p(i) is the probability of item i
            % x(i) is the value of item i
            n = length(p);
            if nargin<2 %~exist('x','var')
                x=1:n;
            end
            self@DiscreteDistribution('DiscreteSampler',3,[min(x),max(x)]);
            setParam(self, 1, 'p', p(:)');
            setParam(self, 2, 'x', x(:)');
            setParam(self, 3, 'f', cumsum(p(:)')/sum(p));
        end
        
        function ex = getMean(self)
            % EX = GETMEAN()

            % Get distribution mean.
            % Weighted by the SUPPORT x, not by the rank 1:n. The two agree on
            % the default support x=1:n, which is why the rank form went
            % unnoticed; on any other support it returned the mean of a
            % distribution the caller never asked for.
            p = self.getParam(1).paramValue;
            x = self.getParam(2).paramValue;
            ex = sum(p(:)'.*x(:)') / sum(p);
        end

        function SCV = getSCV(self)
            % SCV = GETSCV()

            % Get distribution squared coefficient of variation (SCV = variance / mean^2)

            p = self.getParam(1).paramValue;
            x = self.getParam(2).paramValue;
            % `ex` was never assigned here, so every call errored out.
            ex = getMean(self);
            e2 = sum(p(:)'.*(x(:)'.^2)) / sum(p);
            var = e2 - ex^2;
            SCV = var / ex^2;
        end
        
        function X = sample(self, n)
            % X = SAMPLE(N)
            
            x = self.getParam(2).paramValue;
            f = self.getParam(3).paramValue;
            r = rand(n,1);
            X = x(maxpos(min(repmat(r,1,size(f,2))<=f,2)')');
        end
        
        function Ft = evalCDF(self,k)
            % FT = EVALCDF(SELF,K)
            
            f = self.getParam(3).paramValue;
            if k>=1 && k<length(f)
                Ft = f(k);
            else
                Ft = 0;
            end
        end
        
        function pk = evalPMF(self, v)
            % PK = EVALPMF(V)
            
            p = self.getParam(1).paramValue;
            x = self.getParam(2).paramValue;
            if nargin<2 %~exist('v','var')
                pk = p;
                return
            end
            pk = zeros(size(v)); % a point outside the support carries no mass
            for i=1:length(v)
                idx = find(x==v(i),1);
                if ~isempty(idx)
                    pk(i) = p(idx);
                end
            end
        end
        
        function bool = isDisabled(self)
            % BOOL = ISDISABLED()
            
            p = self.getParam(1).paramValue;
            bool = any(isnan(p));
        end
    end
    
end

